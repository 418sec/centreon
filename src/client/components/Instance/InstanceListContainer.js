import React, { Component } from 'react'
import InstanceList from './InstanceList'
import { connect } from 'react-redux'
import { getInstances } from 'Services/hubApi'
import { cleanHeader, replaceTitle} from 'Redux/header/headerActions'

export class InstanceListContainer extends Component {

  constructor (props) {
    super(props)

    this.state = {
      columns: [
        {name: 'companyName', title: 'Company name'},
        {name: 'fingerprint', title: 'Fingerprint'},
        {name: 'serverName', title: 'Server name'},
        {name: 'subscriptionState', title: 'Subscription state'},
        {name: 'unlimitedAccess', title: 'Unlimited access'},
      ],
      currentPage: 0,
      itemsPerPage: 12,
      filters: []
    }
  }

  componentDidMount = () => {
    const { cleanHeader, replaceHeaderTitle } = this.props
    cleanHeader()
    replaceHeaderTitle('Instances')

    this.getData()
  }

  changeFilters = (filters) => {
    this.setState(
      {
        filters: filters,
        currentPage: 0
      },
      this.getData
    )
  }

  changeCurrentPage = (currentPage) => {
    this.setState(
      {currentPage: currentPage},
      this.getData
    )
  }

  /**
   * example : page[number]=1&page[size]=30
   * @returns {string|*}
   */
  getPagination = () => {
    const { currentPage, itemsPerPage } = this.state
    return `?page[number]=${currentPage + 1}&page[size]=${itemsPerPage}`
  }

  /**
   * example : filter[instanceName]=centreon&filter[fingerprint]=abc
   * @returns {string|*}
   */
  getFilters = () => {
    const { filters } = this.state;

    return filters.reduce((acc, { columnName, value }) => {
      acc.push(`filter[${columnName}]=${value}`);
      return acc;
    }, []).join('&');
  }

  getData = () => {
    const { token } = this.props
    this.props.displayInstances(this.getPagination(), this.getFilters(), token)
  }

  render = () => {
    const { columns, currentPage } = this.state
    const { items, itemsPerPage, totalPages, totalItems } = this.props

    const rows = items ?
      Object.keys(items).reduce((acc, key) => {
        acc.push(items[key])
        return acc
      }, [])
      : []

    return (
      <InstanceList
        columns={columns}
        rows={rows}
        itemsPerPage={itemsPerPage}
        totalPages={totalPages}
        totalItems={totalItems}
        currentPage={currentPage}
        changeFilters={this.changeFilters}
        changeCurrentPage={this.changeCurrentPage}
      />
    )
  }
}

const mapStateToProps = (store) => {
  return {
    token: store.authReducer.token,
    items: store.instanceReducer.items,
    totalPages: store.instanceReducer.totalPages,
    totalItems: store.instanceReducer.totalItems,
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    cleanHeader: () => {
      return dispatch(cleanHeader())
    },
    replaceHeaderTitle: (title) => {
      return dispatch(replaceTitle(title))
    },
    displayInstances: (pagination, filters, token) => {
      return dispatch(getInstances(pagination, filters, token))
    },
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(InstanceListContainer)