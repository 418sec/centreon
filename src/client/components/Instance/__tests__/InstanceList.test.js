/* eslint-env jest */
import React from 'react'
import Enzyme from 'enzyme'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'
import { InstanceListContainer as InstanceList} from '../InstanceListContainer'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'

Enzyme.configure({ adapter: new Adapter() })

describe('InstanceListContainer', () => {
  const mockStore = configureStore()

  it('renders correctly', () => {

    let cleanHeader = jest.fn(),
      replaceHeaderTitle = jest.fn(),
      displayInstances = jest.fn()

    const InstanceListComponent = shallow(
      <InstanceList
        cleanHeader={cleanHeader}
        replaceHeaderTitle={replaceHeaderTitle}
        displayInstances={displayInstances}
      />
    ).dive().dive()
    expect(toJson(InstanceListComponent)).toMatchSnapshot()
  })

  describe('filters', () => {

    let cleanHeader = jest.fn(),
      replaceHeaderTitle = jest.fn(),
      displayInstances = jest.fn()

    const items = {
      '1': {
        companyName: 'centreon',
        fingerprint: 'fingerprint',
        serverName: 'centreon-server',
        subscriptionState: {
          instanceId: '1',
          slug: 'instance-unlink-1',
          state: false,
        },
        unlimitedAccess: {
          instanceId: '1',
          serverName: 'centreon-server',
          slug: 'instance-unlimited-1',
          unlimitedMode: false,
        },
      },
      '2': {
        companyName: 'centreon2',
        fingerprint: 'fingerprint2',
        serverName: 'centreon-server2',
        subscriptionState: {
          instanceId: '2',
          slug: 'instance-unlink-2',
          state: false,
        },
        unlimitedAccess: {
          instanceId: '2',
          serverName: 'centreon-server2',
          slug: 'instance-unlimited-2',
          unlimitedMode: false,
        },
      }
    }

    const storeWithItems = mockStore({
      authReducer: {
        token: ''
      },
      instanceReducer: {
        items: items,
        openConfirmDialog: false,
        totalPages: 2,
        totalItems: 4,
      },
      confirmDialogReducer: {
        slug: '',
        allowAction: false,
        open: false,
        content: {
          body: '',
          cancel: '',
          confirm: '',
        }
      }
    })

    const props = {
      token: '',
      items: items,
      totalPages: 2,
      totalItems: 4,
      cleanHeader: cleanHeader,
      replaceHeaderTitle: replaceHeaderTitle,
      displayInstances: displayInstances
    }

    it('Should filter by company name', () => {

      const InstanceListContainer = mount(
        <Provider store={storeWithItems}>
          <InstanceList {...props}/>
        </Provider>
      )

      InstanceListContainer
        .find('th#filterCompanyName input')
        .simulate('change', {target: { value: 'centreo' }})

      expect(InstanceListContainer.childAt(0).instance().state.filters)
        .toEqual([ { columnName: 'companyName', value: 'centreo' } ])
    })

    it('Should filter by fingerprint', () => {

      const InstanceListContainer = mount(
        <Provider store={storeWithItems}>
          <InstanceList {...props}/>
        </Provider>
      )

      InstanceListContainer
        .find('th#filterFingerprint input')
        .simulate('change', {target: { value: 'finger' }})

      expect(InstanceListContainer.childAt(0).instance().state.filters)
        .toEqual([ { columnName: 'fingerprint', value: 'finger' } ])
    })
  })
})