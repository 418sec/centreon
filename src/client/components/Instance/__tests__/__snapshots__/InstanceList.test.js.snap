// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstanceListContainer renders correctly 1`] = `
<div
  className="InstanceList-root-1"
  id="instanceList"
>
  <Grid$$1
    columns={
      Array [
        Object {
          "name": "companyName",
          "title": "Company name",
        },
        Object {
          "name": "fingerprint",
          "title": "Fingerprint",
        },
        Object {
          "name": "serverName",
          "title": "Server name",
        },
        Object {
          "name": "subscriptionState",
          "title": "Subscription state",
        },
        Object {
          "name": "unlimitedAccess",
          "title": "Unlimited access",
        },
      ]
    }
    rows={Array []}
  >
    <FilteringState
      defaultFilters={Array []}
      onFiltersChange={[Function]}
    />
    <PagingState
      currentPage={0}
      defaultCurrentPage={0}
      defaultPageSize={10}
      onCurrentPageChange={[Function]}
    />
    <CustomPaging
      totalCount={0}
    />
    <Table$$1
      cellComponent={[Function]}
      messages={Object {}}
      rowComponent={[Function]}
    />
    <TableHeaderRow$$1
      messages={Object {}}
    />
    <PagingPanel$$1
      messages={Object {}}
    />
    <TableFilterRow$$1
      cellComponent={[Function]}
      messages={Object {}}
    />
  </Grid$$1>
</div>
`;
